### PR TITLE
[main][feat] delta vault to support executor

### PR DESCRIPTION
**Pull request by jakeal-papaca**
## Description
 - delta vault to support executor

## ChangeLogs:
 - DeltaNeutralVault03
 - DeltaNeutralVaultConfig03

### CODE REVIEW

File: solidity/contracts/8.10/protocol/DeltaNeutralVaultConfig03.sol [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+  );
+  event LogSetFeeTreasury(
+    address indexed _caller,
+    address _depositFeeTreasury,
+    address _withdrawFeeTreasury,
+    address _managementFeeTreasury
+  );
+  event LogSetSwapRouter(address indexed _caller, address _swapRouter);
+  event LogSetReinvestPath(address indexed _caller, address[] _reinvestPath);
+  event LogSetExcutor(
```
Comment (Alpaca-Cappuccino) [inline solidity/contracts/8.10/protocol/DeltaNeutralVaultConfig03.sol:60]: LogSetExcutor >>> LogSetExecutor

File: solidity/contracts/8.10/protocol/DeltaNeutralVaultConfig03.sol [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+  error DeltaNeutralVaultConfig_TooMuchBounty(uint256 _bounty);
+  error DeltaNeutralVaultConfig_InvalidSwapRouter();
+  error DeltaNeutralVaultConfig_InvalidReinvestPath();
+  error DeltaNeutralVaultConfig_InvalidReinvestPathLength();
+
+  /// @notice Constants
+  uint8 private constant MIN_LEVERAGE_LEVEL = 3;
+  uint256 private constant MAX_DEPOSIT_FEE_BPS = 1000;
+  uint256 private constant MAX_WITHDRAWAL_FEE_BPS = 1000;
+  uint256 private constant MAX_MANGEMENT_FEE_PER_SEC = 3170979198;
```
Comment (Alpaca-Cappuccino) [inline solidity/contracts/8.10/protocol/DeltaNeutralVaultConfig03.sol:87]: MANAGEMENT

File: solidity/contracts/8.10/protocol/DeltaNeutralVaultConfig03.sol [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+  function isExecutor(address _caller) external view returns (bool) {
+    if (
+      _caller == depositExecutor ||
+      _caller == withdrawExecutor ||
+      _caller == rebalanceExecutor ||
+      _caller == reinvestExecutor
+    ) {
+      return true;
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (pana100x) [reply solidity/contracts/8.10/protocol/DeltaNeutralVaultConfig03.sol:410]: 👍

File: solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+    PositionInfo memory _positionInfoBefore = positionInfo();
+    Outstanding memory _outstandingBefore = _outstanding();
+    _outstandingBefore.nativeAmount = _outstandingBefore.nativeAmount - msg.value;
+
+    // 1. transfer tokens from user to vault
+    _transferTokenToVault(stableToken, _stableTokenAmount);
+    _transferTokenToVault(assetToken, _assetTokenAmount);
+
+    // 2. call execute to do more work.
+    // Perform the actual work, using a new scope to avoid stack-too-deep errors.
```
Comment (alpanaca) [inline solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol:305]: Update this comment to the new implementation

File: solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+      (_lesserPosition.stablePositionDebtValue + _lesserPosition.assetPositionDebtValue);
+
+    return _lpToValue(_lpChange) - _debtChange;
+  }
+
+  /// @notice Proxy function for calling internal action.
+  /// @param _actions List of actions to execute.
+  /// @param _values Native token amount.
+  /// @param _datas The calldata to pass along for more working context.
+  function _execute(
```
Comment (alpanaca) [inline solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol:794]: We don't use this anymore, do we?

**Review by alpanaca — CHANGES_REQUESTED** [review]

File: solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol [hunk 6]
```diff
@@ -0,0 +1,876 @@
...
+    uint256 _withdrawShareValue = shareToValue(_shareToWithdraw);
+
+    // burn shares from share owner
+    _burn(msg.sender, _shareAmount);
+
+    // mint shares equal to withdrawal fee to treasury.
+    _mint(config.withdrawalFeeTreasury(), _shareAmount - _shareToWithdraw);
+
+    // withdraw executor exec
+    IExecutor(config.withdrawExecutor()).exec(_data);
```
Comment (almost-llamas) [inline solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol:358]: The executor needs the withdrawing share amount. I suggest packed it into a single byte.
```
    IExecutor(config.withdrawExecutor()).exec(bytes.concat(abi.encode(_shareToWithdraw), _data));

**Review by almost-llamas — CHANGES_REQUESTED** [review]

File: solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol [hunk 7]
```diff
@@ -0,0 +1,876 @@
...
+    PositionInfo memory _positionInfoBefore = positionInfo();
+    Outstanding memory _outstandingBefore = _outstanding();
+    _outstandingBefore.nativeAmount = _outstandingBefore.nativeAmount - msg.value;
+
+    // 1. transfer tokens from user to vault
+    _transferTokenToVault(stableToken, _stableTokenAmount);
+    _transferTokenToVault(assetToken, _assetTokenAmount);
+
+    // 2. deposit executor exec
+    IExecutor(config.depositExecutor()).exec(_data);
```
Comment (almost-llamas) [inline solidity/contracts/8.10/protocol/DeltaNeutralVault03.sol:305]: Same situation as withdraw function.
```
    IExecutor(config.withdrawExecutor()).exec(bytes.concat(abi.encode(_stableTokenAmount, _assetTokenAmount), _data));

**Review by ricardo-alpaca — APPROVED** [review]

**Review by almost-llamas — APPROVED** [review]

**Review by alpanaca — APPROVED** [review]

**Review by spicysquid168 — APPROVED** [review]

**Pull request merged**